cocotb/types/range.py

```diff
@@ -0,0 +1,175 @@
+from typing import Any, Iterator, overload
+from collections.abc import Sequence
+from functools import cached_property, lru_cache
+
+
+class Range(Sequence):
+    r"""
+    Variant of :class:`range` with inclusive right bound.
+
+    In Python, :class:`range` and :class:`slice` have a non-inclusive right bound.
+    In both Verilog and VHDL, ranges and arrays have an inclusive right bound.
+    This type mimic's Python's :class:`range` type, but supports HDL-like inclusive right bounds.
+    Also supports :attr:`left`, :attr:`right`, and :attr:`length` attributes as seen in VHDL.
+
+    .. code-block:: python3
+
+        >>> r = Range(-2, 3)
+        >>> r.left, r.right, r.length
+        (-2, 3, 6)
+
+        >>> s = Range(8, 'downto', 1)
+        >>> s.left, s.right, s.length
+        (8, 1, 8)
+
+    :meth:`from_range` and :meth:`to_range` can be used to convert from and to :class:`range`.
+
+    .. code-block:: python3
+
+        >>> r = Range(-2, 3)
+        >>> r.to_range()
+        range(-2, 4)
+
+    Supports "null" ranges as seen in VHDL.
+    "null" ranges occur when a left bound cannot reach a right bound with the given direction.
+    They have a length of 0, but the left, right, and direction values remain as given.
+
+    .. code-block:: python3
+
+        >>> r = Range(1, 'to', 0)  # no way to count from 1 'to' 0
+        >>> r.left, r.direction, r.right
+        (1, 'to', 0)
+        >>> r.length
+        0
+
+    .. note::
+        This is only possible when specifying the direction.
+
+    Ranges also support all the features of :class:`range` including, but not limited to:
+
+    - ``value in range`` to see if a vlaue is in the range,
+    - ``range.index(value)`` to see what position in the range the value is,
+    - ``len(range)`` which is equivalent to :attr:`length`.
+
+    The typical use case of this type is in conjunction with :class:`~cocotb.types.Array`.
+
+    Args
+        left: leftmost bound of range
+        direction: 'to' if values are ascending or 'downto' if decending
+        right: rightmost bound of range (inclusive)
+    """
+
+    __slots__ = ("_range", "__dict__")  # __dict__ necessary for cached_property
+
+    @overload
+    def __init__(self, left: int, right: int):
+        pass  # pragma: no cover
+
+    @overload
+    def __init__(self, *, left: int, right: int):
+        pass  # pragma: no cover
+
+    @overload
+    def __init__(self, left: int, direction: str, right: int):
+        pass  # pragma: no cover
+
+    def __init__(self, left, direction=None, right=None):
+        if direction is not None and right is None:
+            right, direction = direction, None
+        if direction is None:
+            step = 1 if left < right else -1
+        else:
+            step = self._translate_direction(direction)
+        self._range = range(left, right + step, step)
+
+    @staticmethod
+    @lru_cache(maxsize=None)
```

Comment:
Do you intend to switch to `@cache` from `_py_compat` of that Logic/Bit PR once it is merged?

Reply:
I can.

Reply:
I moved these changes into this PR as well, and provided a compat layer for `cached_property`.